File: src/security/securityHeaders.ts
/**
 * Security Headers Configuration for Production Deployment
 * Implements comprehensive security headers for web application protection
 */

export interface SecurityHeadersConfig {
  contentSecurityPolicy: string;
  strictTransportSecurity: string;
  xContentTypeOptions: string;
  xFrameOptions: string;
  xXSSProtection: string;
  referrerPolicy: string;
  permissionsPolicy: string;
}

/**
 * Generate Content Security Policy for different environments
 */
export const generateCSP = (environment: 'development' | 'staging' | 'production'): string => {
  const baseCSP = {
    'default-src': ["'self'"],
    'script-src': [
      "'self'",
      environment === 'development' ? "'unsafe-inline'" : '',
      environment === 'development' ? "'unsafe-eval'" : '',
      'https://api.mapbox.com',
      'https://js.sentry-cdn.com',
    ].filter(Boolean),
    'style-src': [
      "'self'",
      "'unsafe-inline'", // Required for Tailwind and component styles
      'https://fonts.googleapis.com',
      'https://api.mapbox.com',
    ],
    'img-src': [
      "'self'",
      'data:',
      'https:',
      '*.supabase.co',
      'https://api.mapbox.com',
      'https://upload.wikimedia.org', // For Wikipedia images
      'https://via.placeholder.com', // For placeholder images
    ],
    'connect-src': [
      "'self'",
      'https://*.supabase.co',
      'https://api.mapbox.com',
      'https://events.mapbox.com',
      'https://api.openweathermap.org',
      'https://en.wikipedia.org',
      environment === 'staging' ? 'https://wheels-wins-backend-staging.onrender.com' : '',
      environment === 'production' ? 'https://pam-backend.onrender.com' : '',
      environment === 'development' ? 'ws://localhost:*' : '',
      environment === 'development' ? 'http://localhost:*' : '',
    ].filter(Boolean),
    'font-src': [
      "'self'",
      'https://fonts.gstatic.com',
    ],
    'media-src': ["'self'"],
    'object-src': ["'none'"],
    'base-uri': ["'self'"],
    'form-action': ["'self'"],
    'frame-ancestors': ["'none'"],
    'upgrade-insecure-requests': environment !== 'development' ? [''] : [],
  };

  return Object.entries(baseCSP)
    .filter(([, values]) => values.length > 0)
    .map(([directive, values]) => `${directive} ${values.join(' ')}`)
    .join('; ');
};

/**
 * Complete security headers configuration
 */
export const getSecurityHeaders = (environment: 'development' | 'staging' | 'production'): SecurityHeadersConfig => ({
  contentSecurityPolicy: generateCSP(environment),

  // HSTS - Force HTTPS for 1 year
  strictTransportSecurity: 'max-age=31536000; includeSubDomains; preload',

  // Prevent MIME type sniffing
  xContentTypeOptions: 'nosniff',

  // Prevent embedding in frames (clickjacking protection)
  xFrameOptions: 'DENY',

  // Enable XSS filtering
  xXSSProtection: '1; mode=block',

  // Control referrer information
  referrerPolicy: 'strict-origin-when-cross-origin',

  // Permissions Policy (formerly Feature Policy)
  permissionsPolicy: [
    'geolocation=(self)',
    'microphone=()',
    'camera=()',
    'magnetometer=()',
    'gyroscope=()',
    'speaker=(self)',
    'notifications=(self)',
    'push=(self)',
  ].join(', ')
});

/**
 * Convert security headers to HTTP header format
 */
export const formatHeadersForHTTP = (headers: SecurityHeadersConfig): Record<string, string> => ({
  'Content-Security-Policy': headers.contentSecurityPolicy,
  'Strict-Transport-Security': headers.strictTransportSecurity,
  'X-Content-Type-Options': headers.xContentTypeOptions,
  'X-Frame-Options': headers.xFrameOptions,
  'X-XSS-Protection': headers.xXSSProtection,
  'Referrer-Policy': headers.referrerPolicy,
  'Permissions-Policy': headers.permissionsPolicy,
});

/**
 * Netlify-specific headers configuration for _headers file
 */
export const generateNetlifyHeaders = (environment: 'staging' | 'production'): string => {
  const headers = getSecurityHeaders(environment);
  const httpHeaders = formatHeadersForHTTP(headers);

  const netlifyConfig = [
    '/*',
    ...Object.entries(httpHeaders).map(([key, value]) => `  ${key}: ${value}`),
    '  Cache-Control: public, max-age=0, must-revalidate',
    '  X-Robots-Tag: index, follow',
    '',
    '# API routes - no caching',
    '/api/*',
    '  Cache-Control: no-store, no-cache, must-revalidate',
    '',
    '# Static assets - long cache',
    '/assets/*',
    '  Cache-Control: public, max-age=31536000, immutable',
    '',
    '# Service worker',
    '/sw.js',
    '  Cache-Control: no-cache',
  ];

  return netlifyConfig.join('\n');
};

/**
 * Validate CSP by checking for common issues
 */
export const validateCSP = (csp: string): { valid: boolean; issues: string[] } => {
  const issues: string[] = [];

  // Check for unsafe practices
  if (csp.includes("'unsafe-eval'")) {
    issues.push("CSP contains 'unsafe-eval' which can allow code injection");
  }

  if (csp.includes('*') && !csp.includes('*.supabase.co')) {
    issues.push("CSP contains wildcard (*) which reduces security");
  }

  // Check for required directives
  const requiredDirectives = ['default-src', 'script-src', 'style-src', 'img-src'];
  requiredDirectives.forEach(directive => {
    if (!csp.includes(directive)) {
      issues.push(`CSP missing required directive: ${directive}`);
    }
  });

  return {
    valid: issues.length === 0,
    issues
  };
};

/**
 * Runtime CSP injection for SPA applications
 */
export const injectCSPMeta = (environment: 'development' | 'staging' | 'production'): void => {
  if (typeof document === 'undefined') return;

  const existingMeta = document.querySelector('meta[http-equiv="Content-Security-Policy"]');
  if (existingMeta) {
    existingMeta.remove();
  }

  const meta = document.createElement('meta');
  meta.httpEquiv = 'Content-Security-Policy';
  meta.content = generateCSP(environment);

  document.head.appendChild(meta);

  console.debug('CSP injected:', meta.content);
};

// Export default configuration
export default {
  getSecurityHeaders,
  generateCSP,
  formatHeadersForHTTP,
  generateNetlifyHeaders,
  validateCSP,
  injectCSPMeta
};

File: backend/app/middleware/security_headers.py
"""
Security Headers Middleware
Adds comprehensive security headers to all responses for protection against common attacks.

Headers Added:
- Content-Security-Policy (CSP): Prevents XSS attacks
- X-Content-Type-Options: Prevents MIME-sniffing
- X-Frame-Options: Prevents clickjacking
- X-XSS-Protection: Legacy XSS protection
- Strict-Transport-Security (HSTS): Forces HTTPS
- Referrer-Policy: Controls referrer information
- Permissions-Policy: Controls browser features

Date: January 10, 2025
"""

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response
from typing import Callable
import secrets


class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """
    Middleware to add security headers to all responses

    CRITICAL SECURITY FIX (Week 2 Thursday):
    - Removed 'unsafe-inline' from CSP
    - Implemented nonce-based CSP for inline scripts/styles
    - Nonce is generated per-request for maximum security
    """

    def __init__(self, app, environment: str = "production"):
        super().__init__(app)
        self.environment = environment
        # Base CSP directives built at init, nonce added per-request
        self._build_base_directives()

    def _build_base_directives(self) -> None:
        """Build base CSP directives (nonce added per-request)"""

        # Base directives shared across environments
        self.base_csp_directives = [
            "default-src 'self'",
            "font-src 'self' https://fonts.gstatic.com data:",
            "img-src 'self' data: https: blob:",
            "frame-src 'self'",
            "object-src 'none'",
            "base-uri 'self'",
            "form-action 'self'",
            "frame-ancestors 'none'",
        ]

        # Environment-specific directives
        if self.environment == "production":
            self.base_csp_directives.extend([
                "connect-src 'self' https://api.openai.com https://generativelanguage.googleapis.com https://*.supabase.co wss://*.supabase.co https://api.mapbox.com https://api.stripe.com",
                "upgrade-insecure-requests",
                "block-all-mixed-content"
            ])
        else:
            # Development/staging: more permissive connect-src
            self.base_csp_directives.extend([
                "connect-src 'self' https: wss: ws:",
                "report-uri /api/security/csp-report"
            ])

        # Enhanced core security headers (same for all environments, no CSP here)
        self.security_headers = {
            # Prevent MIME-sniffing
            "X-Content-Type-Options": "nosniff",

            # Prevent clickjacking
            "X-Frame-Options": "DENY",

            # Legacy XSS protection (still useful for older browsers)
            "X-XSS-Protection": "1; mode=block",

            # Control referrer information
            "Referrer-Policy": "strict-origin-when-cross-origin",

            # Enhanced browser feature control
            "Permissions-Policy": "geolocation=(self), microphone=(self), camera=(self), payment=(self), usb=(), bluetooth=(), magnetometer=(), gyroscope=(), accelerometer=()",

            # Prevent DNS prefetching leaks
            "X-DNS-Prefetch-Control": "off",

            # Prevent MIME confusion attacks
            "X-Download-Options": "noopen",

            # Cross-origin policies
            "Cross-Origin-Embedder-Policy": "require-corp",
            "Cross-Origin-Opener-Policy": "same-origin",
            "Cross-Origin-Resource-Policy": "same-origin",

            # Cache control for sensitive content
            "Cache-Control": "no-store, no-cache, must-revalidate, private",
            "Pragma": "no-cache",
            "Expires": "0",

            # Server information hiding
            "Server": "PAM-Backend",
        }

        # HSTS (HTTP Strict Transport Security) - only in production
        if self.environment == "production":
            self.security_headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains; preload"
            # Remove cache control in production for static assets
            del self.security_headers["Cache-Control"]
            del self.security_headers["Pragma"]
            del self.security_headers["Expires"]

    def _build_csp_with_nonce(self, nonce: str) -> str:
        """Build CSP header with nonce for this request"""
        csp_directives = self.base_csp_directives.copy()

        # CRITICAL FIX: Use nonce instead of unsafe-inline
        # This prevents XSS attacks while allowing legitimate inline scripts/styles
        csp_directives.extend([
            f"script-src 'self' 'nonce-{nonce}' https://cdn.jsdelivr.net https://unpkg.com 'strict-dynamic'",
            f"style-src 'self' 'nonce-{nonce}' https://fonts.googleapis.com",
            f"worker-src 'self'",
            f"manifest-src 'self'",
            f"media-src 'self' data: blob:",
            f"child-src 'none'",
            f"frame-ancestors 'none'",
            f"sandbox allow-same-origin allow-scripts allow-forms allow-popups allow-popups-to-escape-sandbox",
        ])

        return "; ".join(csp_directives)

    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        """
        Add security headers to response with per-request CSP nonce

        IMPORTANT: The nonce must be added to inline scripts/styles in HTML:
        <script nonce="{{ csp_nonce }}">...</script>
        <style nonce="{{ csp_nonce }}">...</style>
        """
        # Generate unique nonce for this request
        nonce = secrets.token_urlsafe(16)

        # Store nonce in request state for templates/responses to use
        request.state.csp_nonce = nonce

        # Process the request
        response = await call_next(request)

        # Add static security headers
        for header_name, header_value in self.security_headers.items():
            response.headers[header_name] = header_value

        # Add CSP header with nonce
        response.headers["Content-Security-Policy"] = self._build_csp_with_nonce(nonce)

        return response


def add_security_headers(app, environment: str = "production"):
    """Helper function to add security headers middleware"""
    app.add_middleware(SecurityHeadersMiddleware, environment=environment)
